src/gui/setuserstatusdialogmodel.cpp

```diff
@@ -0,0 +1,433 @@
+{
+    return Theme::instance()->statusAwayImageSource();
+}
+QUrl SetUserStatusDialogModel::dndIcon() const
+{
+    return Theme::instance()->statusDoNotDisturbImageSource();
+}
+QUrl SetUserStatusDialogModel::invisibleIcon() const
+{
+    return Theme::instance()->statusInvisibleImageSource();
+}
+
+UserStatus::OnlineStatus SetUserStatusDialogModel::onlineStatus() const
+{
+    return _userStatus.state();
+}
+
+QString SetUserStatusDialogModel::userStatusMessage() const
+{
+    return _userStatus.message();
+}
+
+void SetUserStatusDialogModel::setUserStatusMessage(const QString &message)
+{
+    _userStatus.setMessage(message);
+    _userStatus.setMessagePredefined(false);
+}
+
+void SetUserStatusDialogModel::setUserStatusEmoji(const QString &emoji)
+{
+    _userStatus.setIcon(emoji);
+    _userStatus.setMessagePredefined(false);
+    emit userStatusChanged();
+}
+
+QString SetUserStatusDialogModel::userStatusEmoji() const
+{
+    return _userStatus.icon();
+}
+
+void SetUserStatusDialogModel::onUserStatusFetched(const UserStatus &userStatus)
+{
+    if (userStatus.state() != UserStatus::OnlineStatus::Offline) {
+        _userStatus.setState(userStatus.state());
+    }
+    _userStatus.setMessage(userStatus.message());
+    _userStatus.setMessagePredefined(userStatus.messagePredefined());
+    _userStatus.setId(userStatus.id());
+    _userStatus.setClearAt(userStatus.clearAt());
+
+    if (!userStatus.icon().isEmpty()) {
+        _userStatus.setIcon(userStatus.icon());
+    }
+
+    emit userStatusChanged();
+    emit onlineStatusChanged();
+    emit clearAtChanged();
+}
+
+Optional<ClearAt> SetUserStatusDialogModel::clearStageTypeToDateTime(ClearStageType type) const
+{
+    switch (type) {
+    case ClearStageType::DontClear:
+        return {};
+
+    case ClearStageType::HalfHour: {
+        ClearAt clearAt;
+        clearAt._type = ClearAtType::Period;
+        clearAt._period = 60 * 30;
+        return clearAt;
+    }
+
+    case ClearStageType::OneHour: {
+        ClearAt clearAt;
+        clearAt._type = ClearAtType::Period;
+        clearAt._period = 60 * 60;
+        return clearAt;
+    }
+
+    case ClearStageType::FourHour: {
+        ClearAt clearAt;
+        clearAt._type = ClearAtType::Period;
+        clearAt._period = 60 * 60 * 4;
+        return clearAt;
+    }
+
+    case ClearStageType::Today: {
+        ClearAt clearAt;
+        clearAt._type = ClearAtType::EndOf;
+        clearAt._endof = "day";
+        return clearAt;
+    }
+
+    case ClearStageType::Week: {
+        ClearAt clearAt;
+        clearAt._type = ClearAtType::EndOf;
+        clearAt._endof = "week";
+        return clearAt;
+    }
+
+    default:
+        Q_UNREACHABLE();
+    }
+}
+
+void SetUserStatusDialogModel::setUserStatus()
+{
+    clearError();
+    _userStatusJob->setUserStatus(_userStatus);
+}
+
+void SetUserStatusDialogModel::clearUserStatus()
+{
+    clearError();
+    _userStatusJob->clearMessage();
```

Comment:
you make the hypothesis that `_userStatusJob` can be empty so you should test it everywhere